### PR TITLE
Problem notifications

### LINKED ISSUE #2080: Notifications for problems
We require a way of notifying users when there is a problem with their environments (as reported by the problems DB)

### PR DESCRIPTION
**Pull request by bomoko**
This PR introduces new notification classes and severity levels to support notifications that only show certain _kinds_ of information regarding problems (i.e. if we only wanted to be notified about critical problems coming into the system)

# Checklist
- [*] Affected Issues have been mentioned in the Closing issues section
- [*] Documentation has been written/updated
- [*] PR title is ready for changelog and subsystem label(s) applied

# Closing issues
closes #2080

### CODE REVIEW

File: node-packages/commons/src/notificationCommons.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  const rawArray = event.split(":");
+  const eventData = {
+    isProblem: rawArray[0] == "problem",
+    eventType: undefined,
+    dataType: undefined,
+    source: undefined,
+    severityLevel: undefined,
+  }
+  if(eventData.isProblem) {
+    [, eventData.eventType, eventData.source, eventData.dataType, eventData.severityLevel] = rawArray;
```
Comment (timclifford) [inline node-packages/commons/src/notificationCommons.ts]: I'm not sure how the event string is defined (but assume its something like `'problem:insert:drutiny:data:CRITICAL'` etc) that gets split into that rawArray. I've not seen a comma at the start of a de-structured array before, but is that meant to be 'eventData.isProblem' or is it ignoring/skipping the first value purpose? I would say add in the isProblem property also anyways perhaps.

For example:

`[eventData.isProblem, eventData.eventType, eventData.source, eventData.dataType, eventData.severityLevel] = rawArray;`

File: services/api/src/resources/notification/resolvers.ts [hunk 2]
```diff
@@ -14,6 +17,14 @@ const notificationTypeToString = R.cond([
   [R.T, R.identity],
 ]);
 
+const notificationContentTypeToString = R.cond([
+  [R.equals('DEPLOYMENT'), R.toLower],
+  [R.equals('PROBLEM'), R.toLower],
+  [R.T, R.identity],
+]);
+
+
```
Comment (timclifford) [inline services/api/src/resources/notification/resolvers.ts]: just some extra new lines

File: services/api/src/resources/notification/resolvers.ts [hunk 3]
```diff
@@ -231,13 +247,14 @@ export const removeNotificationFromProject: ResolverFn = async (
...
 
-  const select = await query(sqlClient, projectSql.selectProjectByName(input));
+  const select = await query(sqlClient, projectSql.selectProjectByName(input.project));
   const project = R.path([0], select) as any;
 
   await hasPermission('project', 'removeNotification', {
     project: project.id,
   });
 
+
```
Comment (timclifford) [inline services/api/src/resources/notification/resolvers.ts]: and here :D

**Review by timclifford — APPROVED** [review]

File: services/api-db/docker-entrypoint-initdb.d/01-migrations.sql [hunk 4]
```diff
@@ -1127,6 +1127,23 @@ CREATE OR REPLACE PROCEDURE
...
+  BEGIN
+    IF NOT EXISTS(
+      SELECT NULL
+      FROM INFORMATION_SCHEMA.COLUMNS
+      WHERE
+        table_name = 'project_notification'
+        AND table_schema = 'infrastructure'
+        AND column_name = 'content_type'
+    ) THEN
+      ALTER TABLE `project_notification` ADD `content_type` ENUM('deployment', 'problem') NOT NULL DEFAULT 'deployment';
```
Comment (bomoko) [inline services/api-db/docker-entrypoint-initdb.d/01-migrations.sql]: Missing content type here

File: services/api/src/resources/notification/sql.ts [hunk 5]
```diff
@@ -47,8 +64,19 @@ export const Sql = {
         'project.id': input,
       })
       .toString(),
+
+  selectProjectByName: (input) => {
```
Comment (bomoko) [inline services/api/src/resources/notification/sql.ts:68]: is this a copy/paste error?

File: services/logs2email/src/readFromRabbitMQ.ts [hunk 6]
```diff
@@ -277,23 +278,44 @@ export async function readFromRabbitMQ (msg: ConsumeMessage, channelWrapperLogs:
...
   }
   catch (error) {
     logger.error(`No Email information found, error: ${error}`)
     return channelWrapperLogs.ack(msg)
   }
+
   projectEmails.forEach(projectEmail => {
     const { emailAddress } = projectEmail;
 
+    const notificationThresholdMet = notificationContentTypeToInt(projectEmail.notificationSeverityThreshold) <= notificationContentTypeToInt(severityLevel);
```
Comment (bomoko) [inline services/logs2email/src/readFromRabbitMQ.ts:316]: Move this into notification commons as a convenience function maybe?

File: node-packages/commons/src/notificationCommons.ts (hunk 1 above)
Comment (bomoko) [reply node-packages/commons/src/notificationCommons.ts]: Thanks Tim. **isProblem** is a boolean, but you'll see that I've updated it to include **eventClass**, which'll make it less obtuse

File: services/api/src/resources/notification/resolvers.ts (hunk 3 above)
Comment (bomoko) [reply services/api/src/resources/notification/resolvers.ts]: fixed (in next commit)

**Pull request merged**